framework/Optimizers/parentSelectors/parentSelectors.py

```diff
@@ -0,0 +1,156 @@
+    @ In, kwargs, dict, dictionary of parameters for this mutation method:
+          fitness, xr.DataArray, fitness of each chromosome (individual) in the population, i.e., np.shape(fitness) = 1 x populationSize
+          variables, list, variable names.
+          nParents, int, number of required parents.
+    @ Out, selectedParents, xr.DataArray, selected parents, i.e. np.shape(selectedParents) = nParents x nGenes.
+  """
+  # Arguments
+  pop = population.copy()
+  fitness = kwargs['fitness'].copy()
+  nParents= kwargs['nParents']
+  # if nparents = population size then do nothing (whole population are parents)
+  if nParents == pop.shape[0]:
+    return population
+  elif nParents > pop.shape[0]:
+    raise IOError('Number of parents is greater than population size')
+  # begin the roulette selection algorithm
+  selectedParent = xr.DataArray(
+        np.zeros((nParents,np.shape(pop)[1])),
+        dims=['chromosome','Gene'],
+        coords={'chromosome':np.arange(nParents),
+                'Gene': kwargs['variables']})
+  # imagine a wheel that is partitioned according to the selection
+  # probabilities
+
+  for i in range(nParents):
+    # set a random pointer
+    roulettePointer = randomUtils.random(dim=1, samples=1)
+    # Rotate the wheel
+
+    # initialize Probability
+    counter = 0
+    selectionProb = fitness.data/np.sum(fitness.data) # Share of the pie (rouletteWheel)
+    sumProb = selectionProb[counter]
+
+    while sumProb < roulettePointer :
+      counter += 1
+      sumProb += selectionProb[counter]
+    selectedParent[i,:] = pop.values[counter,:]
+    pop = np.delete(pop, counter, axis=0)
+    fitness = np.delete(fitness,counter,axis=0)
+  return selectedParent
+
+def tournamentSelection(population,**kwargs):
+  """
+    Tournament Selection mechanism for parent selection
+    @ In, population, xr.DataArray, populations containing all chromosomes (individuals) candidate to be parents, i.e. population.values.shape = populationSize x nGenes.
+    @ In, kwargs, dict, dictionary of parameters for this mutation method:
+          fitness, np.array, fitness of each chromosome (individual) in the population, i.e., np.shape(fitness) = 1 x populationSize
+          nParents, int, number of required parents
+          variables, list, variable names
+    @ Out, newPopulation, xr.DataArray, selected parents,
+  """
+  fitness = kwargs['fitness']
+  nParents= kwargs['nParents']
+  pop = population.copy()
+
+  popSize = population.values.shape[0]
+
+  selectedParent = xr.DataArray(
+      np.zeros((nParents,np.shape(pop)[1])),
+      dims=['chromosome','Gene'],
+      coords={'chromosome':np.arange(nParents),
+              'Gene': kwargs['variables']})
+
+  if nParents >= popSize/2.0:
+    # generate combination of 2 with replacement
+    selectionList = np.random.choice(np.arange(0,popSize), 2*nParents, replace=False)
```

Comment:
@mandd, in the comment here you said with replacement but you had replace = False, and vise versa. Is the comment wrong or is the code wrong?

Reply:
I guess the code is correct but the comments are not.